File: ios/YATATests/TodoTests.swift
//
//  TodoTests.swift
//  YATATests
//
//  Created by Alpár Szotyori on 20.05.24.
//

import XCTest
@testable import YATA

final class TodoTests: XCTestCase {

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testNewTodoHasRandomUniqueId() throws {
        // Given
        let todo1 = Todo()
        let todo2 = Todo()

        // Then
        XCTAssertTrue(todo1.id != todo2.id)
    }

    func testNewTodoHasNoDeadline() throws {
        // Given
        let todo = Todo()

        // Then
        XCTAssertFalse(todo.hasDeadline())
    }

    func testNewTodoHasEmptyTitleAndDescription() throws {
        // Given
        let todo = Todo()

        // Then
        XCTAssertEqual(todo.title, "")
        XCTAssertEqual(todo.description, "")
    }

    func testRemoveDeadline() throws {
        // Given
        let todo = Todo(deadline: Deadline.now())

        // When
        XCTAssertTrue(todo.hasDeadline())
        let todoWithoutDeadline = todo.removeDeadline()

        // Then
        XCTAssertFalse(todoWithoutDeadline.hasDeadline())
    }

    func testIsOverdueWhenDeadlineHasPassed() throws {
        // Given
        let now = Deadline.now()
        let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: now.date)!
        let deadlineYesterday = Deadline(date: yesterday, timeZone: now.timeZone)

        let todo = Todo(deadline: deadlineYesterday)

        // Then
        XCTAssertTrue(todo.deadline.isOverdue())
    }

    func testIsOverdueWhenDeadlineIsNow() throws {
        // Given
        let now = Deadline.now()

        let todo = Todo(deadline: now)

        // Then
        XCTAssertTrue(todo.deadline.isOverdue(now.date))
    }

    func testIsOverdueWhenCurrentTimeZoneIsAheadOfDeadlineTimeZone() throws {
        // Given
        let dateNow = Date.now

        let deadlineTimeZone = TimeZone(identifier: "UTC+1")!
        let deadlineDate = Date(timeInterval: TimeInterval(deadlineTimeZone.secondsFromGMT(for: dateNow)),
                                since: dateNow)

        let todo = Todo(deadline: Deadline(date: deadlineDate, timeZone: deadlineTimeZone))

        // When
        let currentTimeZone = TimeZone(identifier: "UTC+2")!
        let currentDate = Date(timeInterval: TimeInterval(currentTimeZone.secondsFromGMT(for: dateNow)),
                               since: dateNow)

        // Then
        XCTAssertTrue(todo.deadline.isOverdue(currentDate))
    }

}

File: android/app/src/test/java/de/repeatuntil/yata/domain/TodoTests.kt
package de.repeatuntil.yata.domain

import de.repeatuntil.yata.domain.todolist.entities.Todo
import de.repeatuntil.yata.domain.todolist.valueobjects.Deadline
import kotlinx.datetime.Clock
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.minus
import kotlinx.datetime.toInstant
import kotlinx.datetime.toLocalDateTime
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Test

class TodoTests {

    @Test
    fun `new to-do has a random unique id`() {
        // Given
        val todo1 = Todo()
        val todo2 = Todo()

        // Then
        assert(todo1.id != todo2.id)
    }

    @Test
    fun `new to-do has no deadline`() {
        // Given
        val todo = Todo()

        // Then
        assert(!todo.hasDeadline())
    }

    @Test
    fun `new to-do has empty title and description`() {
        // Given
        val todo = Todo()

        // Then
        assertEquals("", todo.title)
        assertEquals("", todo.description)
    }

    @Test
    fun `new to-do is not completed`() {
        // Given
        val todo = Todo()

        // Then
        assertFalse(todo.isCompleted)
    }

    @Test
    fun `remove deadline`() {
        // Given
        val todo = Todo(deadline = Deadline.now())

        // When
        assert(todo.hasDeadline())
        val todoWithoutDeadline = todo.removeDeadline()

        // Then
        assert(!todoWithoutDeadline.hasDeadline())
    }

    @Test
    fun `is overdue when deadline has passed`() {
        // Given
        val now = Deadline.now()
        val dateTimeNow = now.dateTime
        val instantNow = dateTimeNow.toInstant(now.timeZone)
        val dateTimeYesterday = instantNow.minus(1, DateTimeUnit.DAY, now.timeZone)
            .toLocalDateTime(now.timeZone)

        val todo = Todo(deadline = Deadline(dateTimeYesterday, now.timeZone))

        // Then
        assert(todo.deadline.isOverdue())
    }

    @Test
    fun `is overdue when deadline is now`() {
        // Given
        val instant = Clock.System.now()
        val clock = object : Clock {
            override fun now(): Instant {
                return instant
            }
        }

        val todo = Todo(deadline = Deadline.now(clock))

        // Then
        assert(todo.deadline.isOverdue())
    }

    @Test
    fun `is overdue when current time zone is ahead of deadline's time zone`() {
        // Given
        val deadlineTimeZone = TimeZone.of("UTC+1")

        val todo = Todo(
            deadline = Deadline(
                dateTime = Clock.System.now().toLocalDateTime(deadlineTimeZone),
                timeZone = deadlineTimeZone
            )
        )

        // Given
        val currentTimeZone = TimeZone.of("UTC+2")

        // Then
        assert(todo.deadline.isOverdue(currentTimeZone))
    }
}